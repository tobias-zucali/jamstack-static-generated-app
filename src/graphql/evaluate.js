const { graphql } = require('graphql')

const { root, schema } = require('./schema.js')


module.exports = async ({
  query,
  variables,
  operationName,
}) => {
  const result = await graphql(
    schema,
    query,
    root,
    null,
    variables,
    operationName
  )

  return result
}
