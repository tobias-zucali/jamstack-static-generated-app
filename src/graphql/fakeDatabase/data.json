{
  "manufacturer": [
    {
      "id": "acme",
      "name": "Acme Corporation"
    },
    {
      "id": "moe",
      "name": "Mother Earth Inc."
    }
  ],
  "offers": [
    {
      "id": "a",
      "name": "Banana",
      "manufacturer": "moe",
      "type": "fruit",
      "vitamins": 33
    },
    {
      "id": "b",
      "name": "Manner Wafferl",
      "manufacturer": "acme",
      "sugar": 5,
      "type": "candy"
    },
    {
      "id": "c",
      "name": "Apple",
      "manufacturer": "moe",
      "type": "fruit",
      "vitamins": 37
    },
    {
      "id": "d",
      "manufacturer": "acme",
      "name": "Gummi Bear",
      "sugar": 12,
      "type": "candy"
    }
  ]
}
