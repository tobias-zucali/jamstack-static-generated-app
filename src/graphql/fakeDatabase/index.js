
const {
  products: fakeProducts,
  manufacturer: fakeManufacturers,
} = require('./data.json')

exports.getProducts = () => fakeProducts
exports.getProductBySlug = (slug) => fakeProducts.find((entry) => entry.slug === slug)
exports.getProductIndex = (product) => fakeProducts.indexOf(product)
exports.getProductByIndex = (index) => fakeProducts[index]
exports.getProductByIndexLoop = (index) => {
  let positiveIndex = index % fakeProducts.length
  while (positiveIndex < 0) {
    positiveIndex += fakeProducts.length
  }
  return fakeProducts[positiveIndex]
}

exports.getManufacturerBySlug = (slug) => fakeManufacturers.find((entry) => entry.slug === slug)
