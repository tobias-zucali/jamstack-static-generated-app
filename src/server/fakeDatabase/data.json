{
  "manufacturer": [
    {
      "slug": "acme",
      "name": "Acme Corporation"
    },
    {
      "slug": "moe",
      "name": "Mother Earth Inc."
    }
  ],
  "products": [
    {
      "slug": "banana",
      "name": "Banana",
      "manufacturer": "moe",
      "category": "fruit",
      "vitamins": 33
    },
    {
      "slug": "manner_wafferl",
      "name": "Manner Wafferl",
      "manufacturer": "acme",
      "sugar": 5,
      "category": "candy"
    },
    {
      "slug": "apple",
      "name": "Apple",
      "manufacturer": "moe",
      "category": "fruit",
      "vitamins": 37
    },
    {
      "slug": "gummi_bear",
      "manufacturer": "acme",
      "name": "Gummi Bear",
      "sugar": 12,
      "category": "candy"
    },
    {
      "slug": "comfrey",
      "manufacturer": "moe",
      "name": "Comfrey",
      "category": "medicine"
    }
  ],
  "productCategories": [
    {
      "slug": "fruit",
      "name": "Fruit"
    },
    {
      "slug": "candy",
      "name": "Candy"
    },
    {
      "slug": "medicine",
      "name": "Medicine"
    }
  ]
}
