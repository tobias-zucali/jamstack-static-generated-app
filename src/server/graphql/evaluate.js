import { graphql } from 'graphql'

import schema from './schema'


export default async function evaluate({
  query,
  variables,
  operationName,
}) {
  const result = await graphql(
    schema,
    query,
    null,
    null,
    variables,
    operationName
  )

  return result
}
