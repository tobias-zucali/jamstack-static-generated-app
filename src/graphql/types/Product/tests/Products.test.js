import evaluate from '../../../evaluate'

import { productsDB } from '../../../fakeDatabase'


describe('graphql/Products', () => {
  it('returns all products', async () => {
    await expect(
      evaluate({
        query: `
            {
              allProducts {
                products {
                  slug
                  name
                }
                totalCount
                edges {
                  first {
                    slug
                  }
                  last {
                    slug
                  }
                  next {
                    slug
                  }
                  previous {
                    slug
                  }
                }
              }
            }
          `,
      })
    ).resolves.toEqual({
      data: {
        allProducts: {
          products: productsDB.getList().map(({ slug, name }) => ({ slug, name })),
          totalCount: productsDB.getList().length,
          edges: {
            next: null,
            previous: null,
            first: {
              slug: productsDB.getByIndexLoop(0).slug,
            },
            last: {
              slug: productsDB.getByIndexLoop(-1).slug,
            },
          },
        },
      },
    })
  })

  it('returns some products', async () => {
    await expect(
      evaluate({
        query: `
            {
              allProducts(after: "${productsDB.getByIndexLoop(0).slug}", limit: 2) {
                products {
                  slug
                }
                totalCount
                edges {
                  first {
                    slug
                  }
                  last {
                    slug
                  }
                  next {
                    slug
                  }
                  previous {
                    slug
                  }
                }
              }
            }
          `,
      })
    ).resolves.toEqual({
      data: {
        allProducts: {
          products: [
            { slug: productsDB.getByIndexLoop(1).slug },
            { slug: productsDB.getByIndexLoop(2).slug },
          ],
          totalCount: productsDB.getList().length,
          edges: {
            previous: {
              slug: productsDB.getByIndexLoop(0).slug,
            },
            next: {
              slug: productsDB.getByIndexLoop(3).slug,
            },
            first: {
              slug: productsDB.getByIndexLoop(1).slug,
            },
            last: {
              slug: productsDB.getByIndexLoop(2).slug,
            },
          },
        },
      },
    })
  })

  it('returns candy products', async () => {
    await expect(
      evaluate({
        query: `
            {
              allProducts(category: "candy", limit: 1) {
                products {
                  slug
                  category {
                    slug
                  }
                  ...on Candy {
                    sugar(unit: KG)
                  }
                }
              }
            }
          `,
      })
    ).resolves.toEqual({
      data: {
        allProducts: {
          products: [
            {
              slug: productsDB.getByIndexLoop(1).slug,
              sugar: 0.005,
              category: {
                slug: 'candy',
              },
            },
          ],
        },
      },
    })
  })
})
