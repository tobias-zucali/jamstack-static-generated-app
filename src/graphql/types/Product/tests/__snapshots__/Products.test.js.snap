// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql/Products matches all products snapshot 1`] = `
Object {
  "data": Object {
    "allProducts": Object {
      "nodes": Array [
        Object {
          "category": Object {
            "description": null,
            "name": "Fruit",
            "slug": "fruit",
          },
          "manufacturer": Object {
            "address": null,
            "description": null,
            "name": "Mother Earth Inc.",
            "slug": "moe",
          },
          "name": "Banana",
          "slug": "banana",
          "vitamins": 33,
        },
        Object {
          "category": Object {
            "description": null,
            "name": "Candy",
            "slug": "candy",
          },
          "manufacturer": Object {
            "address": null,
            "description": null,
            "name": "Acme Corporation",
            "slug": "acme",
          },
          "name": "Manner Wafferl",
          "slug": "manner_wafferl",
          "sugar": 5,
        },
        Object {
          "category": Object {
            "description": null,
            "name": "Fruit",
            "slug": "fruit",
          },
          "manufacturer": Object {
            "address": null,
            "description": null,
            "name": "Mother Earth Inc.",
            "slug": "moe",
          },
          "name": "Apple",
          "slug": "apple",
          "vitamins": 37,
        },
        Object {
          "category": Object {
            "description": null,
            "name": "Candy",
            "slug": "candy",
          },
          "manufacturer": Object {
            "address": null,
            "description": null,
            "name": "Acme Corporation",
            "slug": "acme",
          },
          "name": "Gummi Bear",
          "slug": "gummi_bear",
          "sugar": 12,
        },
        Object {
          "category": Object {
            "description": null,
            "name": "Medicine",
            "slug": "medicine",
          },
          "manufacturer": Object {
            "address": null,
            "description": null,
            "name": "Mother Earth Inc.",
            "slug": "moe",
          },
          "name": "Comfrey",
          "slug": "comfrey",
        },
      ],
      "totalCount": 5,
    },
  },
}
`;
